### DISCUSSION TITLE
Exploring Oqtane as a Headless CMS with API Security

### DISCUSSION THREAD

**Discussion by Rodien** [oqtane/oqtane.framework#3451] [Q&A]
I want to experiment with using Oqtane as a headless system. As a result, I've created a module that will serve as an API.

Now, I'd like to secure the API using an Identity Provider (IdP).

My goal is to utilize Oqtane as a headless CMS while maintaining a separate project for the front-end.

I've added the following code to my Oqtane module, and I've previously used this code successfully in a prototype application that's not based on Oqtane. However, when I use it within Oqtane and attempt to call the API, I consistently receive a 403 status code.

I believe I must be overlooking something in the configuration or integration. Any insights or guidance would be greatly appreciated.

![image](https://github.com/oqtane/oqtane.framework/assets/50739829/53e8b697-454d-4aff-af56-babf1019da72)

I am using the following console application to run some quick tests:

![image](https://github.com/oqtane/oqtane.framework/assets/50739829/ff6cd0b4-8744-4e56-b24c-e5658c3a1a97)

Is this approach even possible when using Oqtane, @sbwalker ?

**Comment by sbwalker** [comment]
@Rodien so you want to create a separate front-end application... however you failed to mention if the front-end will be a web, desktop, or mobile application.

**Reply by Rodien** [reply]
@sbwalker I will start with web, and after that, desktop and mobile.

I am just trying to see what the possibilities are, so I am creating some quick and simple prototypes.

**Reply by sbwalker** [reply]
I was asking because different approaches should be used for different environments

**Reply by Rodien** [reply]
Yeah, I want to try using a separate Blazor WebAssembly project, but I also want to create a prototype using something like the UNO Platform at some point.

**Comment by sbwalker** [comment]
I would suggest you take a look at:

https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Server/Infrastructure/Middleware/JwtMiddleware.cs

You will see that if a request contains an auth token, it will deal with multi-tenancy to determine the Site, and then validate the token with the Secret, Issuer, Audience defined in User Settings for the Site. It then looks for specific claims so that it can get the corresponding User in Oqtane and load their Roles.
